superset-frontend/src/dashboard/util/permissionUtils.ts

```diff
@@ -58,8 +57,12 @@ export function canUserAccessSqlLab(
   return (
     isUserAdmin(user) ||
     (isUserWithPermissionsAndRoles(user) &&
-      Object.keys(user.roles || {}).some(
-        role => role.toLowerCase() === SQL_LAB_ROLE,
-      ))
+      Object.values(user.roles || {})
+        .flat()
+        .some(
+          permissionView =>
+            permissionView[0] === 'menu_access' &&
```

Comment:
`canUserAccessSqlLab` is called from places where a menu is not present such as the home page. Could you change the code to request the `menu_access` permission at the places where a menu is present?

Reply:
Good catch. Maybe the appropriate util name here would be `canUserAccessSqlLabMenu`, and for the Home page we should probably be checking for the `['can_read' 'SavedQuery']` perm. Thoughts @giftig @michael-s-molina ?

Reply:
We've been kind of iffy on the concept of menu access permissions, and not relying on them/using them at all. Rather if there are multiple actions with different perms, show all of the actions and disable the ones where the user doesn't have access like [this](https://github.com/preset-io/superset/blob/0d2ec229f62b01046947d915f0d7d31d63078dfa/superset-frontend/src/views/CRUD/data/database/DatabaseList.tsx#L176), or if they don't have access to anything in the menu, then don't show the menu at all. What we've been trying to avoid is having menu access, and then the menu being empty because the user had no other access to anything in it. This point is really more of a product question so cc @yousoph or @kasiazjc.

Reply:
+1 on @eschutho's suggestion. It looks way more clear to check for the resources/actions than the menu itself.

Reply:
Yeah that definitely makes sense on home, I see it's only trying to get the saved queries so `menu_access` doesn't seem correct. On `DatasourceControl` the `menu_access` check probably does make sense, though.

So the best way to stay DRY here but use granular checks is probably replace it with a more generic `userHasPermission(user, viewName, permissionName)`; surprised that doesn't exist already actually.

Reply:
Used and unit tested the new util and the build is back to green now. :white_check_mark:

Reply:
> On DatasourceControl the menu_access check probably does make sense, though, because we're displaying a link to SQL Lab, same as how the top menu bar would work.

@giftig Could you elaborate more on this? I was interpreting @eschutho's comment in a way that if a user has the `SQL Lab` permission, all possible ways to get to SQL Lab would be enabled such as menu items, links, etc. In that context, we wouldn't have the `menu_access` permission.

Reply:
@michael-s-molina `SQL Lab` is the view rather than the individual permission, and there are a few different permissions under it, but not sure which other permissions would be more suitable to gate the appearance of this button than `menu_access`. There's actually an added wrinkle that it's `[SQL Lab].[menu_access]` but other permissions are on `SQLLab` instead, but `SQLLAB_ONLY_PERMISSIONS` in the security manager give you a bit of a view of which ones apply:

```python
    SQLLAB_ONLY_PERMISSIONS = {
        ("can_my_queries", "SqlLab"),
        ("can_read", "SavedQuery"),
        ("can_write", "SavedQuery"),
        ("can_export", "SavedQuery"),
        ("can_read", "Query"),
        ("can_export_csv", "Query"),
        ("can_get_results", "SQLLab"),
        ("can_execute_sql_query", "SQLLab"),
        ("can_estimate_query_cost", "SQL Lab"),
        ("can_export_csv", "SQLLab"),
        ("can_sqllab_history", "Superset"),
        ("can_sqllab", "Superset"),
        ("can_test_conn", "Superset"),  # Deprecated permission remove on 3.0.0
        ("can_activate", "TabStateView"),
        ("can_get", "TabStateView"),
        ("can_delete_query", "TabStateView"),
        ("can_post", "TabStateView"),
        ("can_delete", "TabStateView"),
        ("can_put", "TabStateView"),
        ("can_migrate_query", "TabStateView"),
        ("menu_access", "SQL Lab"),
        ("menu_access", "SQL Editor"),
        ("menu_access", "Saved Queries"),
        ("menu_access", "Query Search"),
    }
```

@eschutho's comment referred to displaying a menu and hiding the menu if it would be empty, rather than relying on `menu_access` features, but I'm not sure we can make that apply here. This is gating an individual menu item under the datasource [..] menu, and that menu item is just a link to SQL Lab. `[SQL Lab].[menu_access]` decides whether the SQL Lab link appears in the top bar, so imo it would make sense to follow the same logic here.

If we gate this by something else, the check may have to be quite a bit more complicated -- which permission(s) are required for it to be sensible to let you access SQL Lab? `Superset.can_sqllab` might be an alternative; that one's needed to be able to load the SQL Lab content at all, I believe. You'd probably also need at least `SQLLab.can_execute_sql_query` for it to make sense to be sent there, though, or you wouldn't be able to actually use it.

Is the idea is to deprecate `menu_access` permissions entirely going forwards? If we really want to deprecate `menu_access` as a permission we should probably change the conditions in which the main menu item appears to match whatever the logic should be here, too.

Reply:
Thank you for the additional context @giftig. We should definitely simplify our permission model as our current granularity is way more complex than what's required in practice! I think this type of permission review requires a SIP and holistic thinking so I added a card [here](https://github.com/orgs/apache/projects/201/views/1?pane=issue&itemId=35462560). With that in mind, you comment seems appropriate for the scope of this PR:

> [SQL Lab].[menu_access] decides whether the SQL Lab link appears in the top bar, so imo it would make sense to follow the same logic here.